### ISSUE TITLE
New architecture?

### ISSUE THREAD

**Issue by hartenthaler** [Neriderc/GVExport#27]
I had a look into the source code. As far as I understood there are three steps

1. addIndiToList
2. use two arrays: individuals and families
3. visualize

Maybe you have seen the custom modules ["ClippingsCart - Extended"](https://github.com/huhwt/huhwt-cce), and the documentation with a [system architecture diagram](https://wiki.genealogy.net/Webtrees_Handbuch/Anleitung_für_Webmaster/Erweiterungsmodule/Sammelbehälter). And there are two new visualization modules: [TAM](https://github.com/huhwt/huhwt-wttam) and [Lineage](https://github.com/huhwt/huhwt-wtlin).
The idea behind this architecture is

- use several modules as a source in adding things to the clippings cart
- use functions to manage the clippings cart (for example delete records)
- use function modules like export of GEDCOM or visualize the data (GVExport, TAM, Lineage, ...)

My question is now if it would make sense to use GVExport as the third module beside TAM and Lineage and use the clippings cart as a source of individuals and families. The advantage would be that the clippings cart is much more flexible than the existing addIndiToList algorithm. You can use many functions to populate the clippings cart with individuals and families and then visualize them by several tools, like GVExport, TAM, or Lineage.

In the first step, I would try to write a function for GVExport that reads the content of the clippings cart and produces the two arrays "individuals" and "families", so that addIndiToList can be replaced by this new function or can be used in parallel to the new function. This would allow some testing whether the visualizing part of GVExport is stable enough to work with data from other sources than the addIndiToList function.

Because the functionality of addIndiToList is very helpful to get a diagram of the persons and families next to a person, I would provide in the second step a new menu topic for the "clippings cart extended" that provides the same functionality as it is used in GVExport at the moment (addIndiToList). This can be combined with an old webtrees 1 module "branch export" that uses a starting person and several stop persons/families (stored as profile), which is similar to GVExport. This old module can be recycled.

As a summary
- advantage for users: more flexibility in selecting persons/families to be visualized and harmonized concept how to visualize larger sets of records (GVExport, TAM, Lineage)
- advantage for developers: more modular architecture (GVExport concentrates on visualization instead of additionally collecting siblings, spouses, cousins, ...)
- disadvantage: greater complexity, because you need the clippings cart (that is disabled today in many webtrees installations) and two steps in two different environments to select the individuals and to visualize them instead of doing these two steps in only one environment (as it is done in GVExport today)

**Comment by Neriderc** [comment]
I have a few questions about this approach. Firstly, I'm not sure I understand the advantage to the users. I haven't used those other modules, perhaps if I had I would understand, would the main feature be that users could go to different individuals pages and add them to the clippings cart, then go to the module and have them all displayed in GVExport?

Secondly, would this rely on having the extended clipping cart installed? i.e. would users have to have ClippingsCart - Extended installed in order to use GVExport?

Lastly, can we retain the ability to select the starting individuals from the GVExport menu under this system by  keeping the person search, and having them added to the clippings cart when chosen?

**Comment by hartenthaler** [comment]
Yes, you could go to several individuals, and add them and their ancestors or descendants or their families to the clippings cart. Then you can delete some persons in the clippings cart if necessary. Finally, you can show them all using GVExport. The clippings cart exists for every tree during a browser session.

No, this is working with the standard clippings cart, too. But using the extended clippings cart adds more functionality. You can use the standard clippings cart in parallel with the Vesta clippings cart and the "extended clippings cart" or you can activate only the standard clippings cart or only the extended clippings cart. If you like to use the new modules TAM or Lineage you have to use the extended clippings cart, because they can be started in this module only.

Your "lastly" idea should be easy to implement. If you have the arrays "individuals" and "families" filled then it is no problem to copy them to the clippings cart, because the clippings cart is nothing else than an array of XREFs.

**Comment by Neriderc** [comment]
I don't have a problem with this idea in theory, but I would like to keep existing functionality from a user perspective. So if you add a person in the GVExport selection box, and select to add all cousins, siblings, spouses, and choose X generations of ancestors or descendants, this should be honoured.

If we can maintain existing functionality, but also support loading the individuals and families from the clippings cart, then I'd support that. Is the clippings cart commonly disabled in webtrees installations? An alternative option could be to keep the existing arrays but populate them from the clipping cart if there are records to load. Or alternatively, load all the PIDs from the clippings cart into the PID box in GVExport, then run using existing codebase ( I haven't tested the maximum size of this, which could be an issue). 

I guess it depends on what your main objective is. Are you concerned at the current architecture and wanting this tidier, or are you more interested in the functionality that this change would bring?

**Comment by Neriderc** [comment]
In my experience the clippings cart is enabled in a default installation, certainly it was for me. Unless we know of large numbers of people disabling it, I'd think it would be ok to require a built-in module for GVExport to work.

I'm happy for you to attempt this, keeping in mind we want to maintain existing functionality.
